conda_build/index/__init__.py

```diff
@@ -921,25 +1089,305 @@ def index_subdir(self, subdir, index_file=None, verbose=False, progress=False):
+                },
+                "repodata_version": REPODATA_VERSION,
+                "removed": sorted(list(ignore_set)),
+            }
+        finally:
+            if stat_cache != stat_cache_original:
+                cache.save_stat_cache(stat_cache)
+        return new_repodata
+
+    def index_subdir_sql(self, subdir, index_file=None, verbose=False, progress=False):
+        subdir_path = join(self.channel_root, subdir)
+
+        cache = sqlitecache.CondaIndexCache(
+            channel_root=self.channel_root, channel=self.channel_name, subdir=subdir
+        )
+        if cache.cache_is_brand_new:
+            # guaranteed to be only thread doing this?
+            cache.convert()
+
+        repodata_json_path = join(subdir_path, REPODATA_FROM_PKGS_JSON_FN)
+
+        if verbose:
+            log.info("Building repodata for %s" % subdir_path)
+
+        # gather conda package filenames in subdir
+        # we'll process these first, because reading their metadata is much faster
+        # XXX eliminate all listdir. that and stat calls are significant.
+        log.info("listdir")
+        fns_in_subdir = {
+            fn for fn in os.listdir(subdir_path) if fn.endswith((".conda", ".tar.bz2"))
+        }
+
+        # load current/old repodata XXX with sql cache, where 'select from
+        # index_json' is incredibly fast, this code might be easier to reason
+        # about without carrying over old repodata.packages
+        try:
+            with open(repodata_json_path) as fh:
+                old_repodata = json.load(fh) or {}
+        except (OSError, JSONDecodeError):
+            # log.info("no repodata found at %s", repodata_json_path)
+            old_repodata = {}
+
+        # could go in a temporary table
+        old_repodata_packages = old_repodata.get("packages", {})
+        old_repodata_conda_packages = old_repodata.get("packages.conda", {})
+        old_repodata_fns = set(old_repodata_packages) | set(old_repodata_conda_packages)
+
+        # put filesystem 'ground truth' into stat table
+        # will we eventually stat everything on fs, or can we shortcut for new?
+        def listdir_stat():
+            for fn in fns_in_subdir:
+                abs_fn = os.path.join(subdir_path, fn)
+                stat = os.stat(abs_fn)
+                yield {
+                    "path": cache.database_path(fn),
+                    "mtime": int(stat.st_mtime),
+                    "size": stat.st_size,
+                }
+
+        log.info("save fs state")
+        with cache.db:
+            # XXX and path like channel/% if multiple channels share cache
+            cache.db.execute("DELETE FROM stat WHERE stage='fs'")
+            cache.db.executemany(
+                """
+            INSERT INTO STAT (stage, path, mtime, size)
+            VALUES ('fs', :path, :mtime, :size)
+            """,
+                listdir_stat(),
+            )
+
+        log.info("repodata into stat")
```

Comment:
probably continue doing the repodata compare in memory

Reply:
the part where we have to compare with fs to load the cache, and with repodata is confusing.

could we split into two functions: 1. make sure cache knows about everything in repodata 2. load from cache into repodata as necessary.